localstack/services/cloudwatch/provider_v2.py

```diff
@@ -109,6 +158,72 @@ def delete_alarms(self, context: RequestContext, alarm_names: AlarmNames) -> Non
             alarm_arn = arns.cloudwatch_alarm_arn(alarm_name)  # obtain alarm ARN from alarm name
             self.alarm_scheduler.delete_scheduler_for_alarm(alarm_arn)
 
+    def put_metric_data(
+        self, context: RequestContext, namespace: Namespace, metric_data: MetricData
+    ) -> None:
+        _validate_parameters_for_put_metric_data(metric_data)
+
+        self.cloudwatch_database.add_metric_data(
+            context.account_id, context.region, namespace, metric_data
+        )
+
+    def get_metric_data(
+        self,
+        context: RequestContext,
+        metric_data_queries: MetricDataQueries,
+        start_time: Timestamp,
+        end_time: Timestamp,
+        next_token: NextToken = None,
+        scan_by: ScanBy = None,
+        max_datapoints: GetMetricDataMaxDatapoints = None,
+        label_options: LabelOptions = None,
+    ) -> GetMetricDataOutput:
+        results: List[MetricDataResults] = []
+        limit = max_datapoints or 100_800
+        messages: MetricDataResultMessages = []
+        nxt = None
+        for query in metric_data_queries:
+            query_result = self.cloudwatch_database.get_metric_data_stat(
+                account_id=context.account_id,
+                region=context.region,
+                query=query,
+                start_time=start_time,
+                end_time=end_time,
+                scan_by=scan_by,
+            )
+            if query_result.get("messages"):
+                messages.extend(query_result.get("messages"))
+
+            # Paginate
+            aliases_list = PaginatedList(query_result.get("datapoints", {}).items())
+            page, nxt = aliases_list.get_page(
+                lambda metric_result: metric_result.get("Id"),
+                next_token=next_token,
+                page_size=limit,
+            )
```

Comment:
are we sure that the the pagination is for the individual datapoints, and not for the entire "formatted_results"? 🤔